File: 351004/Shkredova/discussion/src/main/java/com/example/discussion/controller/NoticeController.java
package com.example.discussion.controller;

import com.example.discussion.dto.NoticeRequestTo;
import com.example.discussion.dto.NoticeResponseTo;
import com.example.discussion.service.NoticeService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/api/v1.0/notices")
public class NoticeController {
    private final NoticeService noticeService;

    public NoticeController(NoticeService noticeService) {
        this.noticeService = noticeService;
    }

    @PostMapping
    public ResponseEntity<NoticeResponseTo> create(@Valid @RequestBody NoticeRequestTo request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(noticeService.create(request));
    }

    @GetMapping
    public ResponseEntity<List<NoticeResponseTo>> findAll() {
        return ResponseEntity.ok(noticeService.findAll());
    }

    @GetMapping("/{id}")
    public ResponseEntity<NoticeResponseTo> findById(@PathVariable Long id) {
        return noticeService.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/news/{newsId}")
    public ResponseEntity<List<NoticeResponseTo>> findByNewsId(@PathVariable Long newsId) {
        return ResponseEntity.ok(noticeService.findAllByNewsId(newsId));
    }

    @GetMapping("/news/{newsId}/id/{id}")
    public ResponseEntity<NoticeResponseTo> findByNewsIdAndId(@PathVariable Long newsId, @PathVariable Long id) {
        return noticeService.findByNewsIdAndId(newsId, id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PutMapping
    public ResponseEntity<NoticeResponseTo> update(@Valid @RequestBody NoticeRequestTo request) {
        return ResponseEntity.ok(noticeService.update(request));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        var notice = noticeService.findById(id);
        if (notice.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        noticeService.delete(notice.get().getNewsId(), id);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/news/{newsId}/id/{id}")
    public ResponseEntity<Void> deleteWithNewsId(@PathVariable Long newsId, @PathVariable Long id) {
        noticeService.delete(newsId, id);
        return ResponseEntity.noContent().build();
    }
}

File: 351003/brezgunov/discussion/src/main/kotlin/com/example/discussion/controller/NoticeController.kt
package com.example.discussion.controller

import com.example.discussion.dto.ErrorResponseTo
import com.example.discussion.dto.notice.NoticeRequestTo
import com.example.discussion.dto.notice.NoticeResponseTo
import com.example.discussion.exception.AbstractException
import com.example.discussion.service.NoticeService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/{version}/notices")
class NoticeController(
    private val noticeService: NoticeService
) {
    @ExceptionHandler(AbstractException::class)
    fun handleUserNotFound(e: AbstractException): ResponseEntity<ErrorResponseTo> {
        val error = ErrorResponseTo(
            status = e.errorCode.value(),
            message = e.errorMsg
        )
        return ResponseEntity.status(e.errorCode).body(error)
    }

    @PostMapping(version = "1.0")
    @ResponseStatus(HttpStatus.CREATED)
    fun createNotice(@Valid @RequestBody noticeRequestTo: NoticeRequestTo): NoticeResponseTo {
        return noticeService.createNotice(noticeRequestTo)
    }

    @GetMapping(path = ["/{id}"], version = "1.0")
    fun readNoticeById(@PathVariable id: Long): NoticeResponseTo {
        return noticeService.readNoticeById(id)
    }

    @GetMapping(version = "1.0")
    fun readAllnotices(): List<NoticeResponseTo> = noticeService.readAllNotices()

    @PutMapping(path = ["/{id}", ""], version = "1.0")
    @ResponseStatus(HttpStatus.OK)
    fun updateNotice(
        @Valid @RequestBody noticeRequestTo: NoticeRequestTo,
        @PathVariable("id") noticeId: Long
    ) =
        noticeService.updateNotice(noticeId, noticeRequestTo)

    @DeleteMapping(path = ["/{id}"], version = "1.0")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteNotice(@PathVariable id: Long) {
        noticeService.removeNoticeById(id)
    }
}